### PR TITLE
Make assert_script_sudo work properly in serial terminal

### PR DESCRIPTION
**Pull request by b10n1k**
When `is_serial_terminal` is used the output doesnt need redirection to the serial device. This causes the assertion to fail when the serial terminal is used.

Signed-off-by: ybonatakis <ybonatakis@suse.com>

### CODE REVIEW

**Comment by b10n1k** [comment]
i know that this needs tests. i didnt find any for `assert_script_sudo` in additional. i am working on that

File: testapi.pm [hunk 1]
```diff
@@ -1030,7 +1030,9 @@ C<$serialdev>.
 sub assert_script_sudo {    # no:style:signatures
     my ($cmd, $wait) = @_;
     my $str = hashed_string("ASS$cmd");
-    script_sudo("$cmd; echo $str-\$?- > /dev/$serialdev", 0);
+    my $redirect_to_serial_console = if is_serial_terminal() ? '' : "> /dev/$serialdev";
+    my $marker = sprintf "echo $str-\$?- %s", $redirect_to_serial_console;
```
Comment (okurz) [inline testapi.pm:1034]: It should suffice to just do

```
my $marker = "echo $str-\$?-$redirect_to_serial_console";
```

**Review by okurz — COMMENTED** [review]
Looks much better then before, I like it :)

File: testapi.pm [hunk 2]
```diff
@@ -1030,7 +1030,9 @@ C<$serialdev>.
 sub assert_script_sudo {    # no:style:signatures
     my ($cmd, $wait) = @_;
     my $str = hashed_string("ASS$cmd");
-    script_sudo("$cmd; echo $str-\$?- > /dev/$serialdev", 0);
+    my $redirect_to_serial_console = if is_serial_terminal() ? '' : "> /dev/$serialdev";
```
Comment (perlpunk) [inline testapi.pm:1033]: The `if` should be removed:
```suggestion
    my $redirect_to_serial_console = is_serial_terminal() ? '' : "> /dev/$serialdev";
```

File: testapi.pm (hunk 1 above)
Comment (perlpunk) [reply testapi.pm:1034]: I would still add a space, though

Comment (perlpunk) [reply testapi.pm:1034]: hm, the space should probably be added to the beginning of `$redirect_to_serial_console` and not at the end of `$marker`, though.

File: testapi.pm (hunk 2 above)
Comment (perlpunk) [reply testapi.pm:1033]: slight correction:
```suggestion
    my $redirect_to_serial_console = is_serial_terminal() ? '' : " > /dev/$serialdev";
```
and then what @okurz suggested.

File: testapi.pm (hunk 1 above)
Comment (b10n1k) [reply testapi.pm:1034]: i am not sure what's the problem with the sprintf? or it is not request (maybe just a personal preference?)

Comment (okurz) [reply testapi.pm:1034]: Well, I assume there is unnecessary overhead when calling sprintf. sprintf might be required if the requirement is higher but here the simple variable expansion within strings work. This is Perl, not C :)

Comment (b10n1k) [reply testapi.pm]: *ready for review*

File: testapi.pm (hunk 2 above)
Comment (b10n1k) [reply testapi.pm]: *ready for review*

**Comment by perlpunk** [comment]
There seems to be a problem with codecov. I retriggered the test (already a second time)

**Review by Martchus — APPROVED** [review]

File: t/03-testapi.t [hunk 3]
```diff
@@ -334,6 +334,21 @@ subtest 'script_run' => sub {
...
 };
 
+sub assert_script_sudo_test ($is_serial_terminal) {
+    my $mock_testapi = Test::MockModule->new('testapi');
+    $mock_testapi->redefine(_handle_found_needle => sub { return $_[0] });
+    $mock_testapi->noop(qw(send_key enter_cmd));
+    $mock_testapi->redefine(hashed_string => 'XXX');
+    $mock_testapi->redefine(wait_serial => "XXX-0-");
+    my $info = $is_serial_terminal ? 'with serial_terminal' : 'without serial_terminal';
+    is(assert_script_sudo('echo foo'), undef, 'sucessfull assertion of script_sudo $info');
```
Comment (perlpunk) [inline t/03-testapi.t:344]: If someone changes the `assert_script_sudo` to what it did before, then the test would still pass, so I would say that the test is not sufficient.
Maybe you can mock `script_sudo` and check the parameter in the mocked function.

Also `s/successfull/successful/` :)

**Review by perlpunk — CHANGES_REQUESTED** [review]
I forgot to request changes when I commented

File: t/03-testapi.t (hunk 3 above)
Comment (kalikiana) [reply t/03-testapi.t]: @b10n1k I see you added mocking. Please don't forget about the typo.

**Comment by perlpunk** [comment]
I hope it's ok that I pushed some minor improvements.
The `assert_script_sudo_test` didn't do anything different regarding the given `$is_serial_terminal` parameter (just did the same test twice with a different test label), so it's now only called once.
And I mocked `script_sudo` to check that it is called like expected.

**Review by Martchus — APPROVED** [review]

File: testapi.pm [hunk 4]
```diff
@@ -1012,6 +1012,12 @@ sub background_script_run {    # no:style:signatures
     return $distri->background_script_run($cmd, %args);
 }
 
+sub _set_assert_marker {
+    my ($hashed_string) = @_;
```
Comment (okurz) [inline testapi.pm:1016]: For any new functions please use signatures, i.e.:

```suggestion
sub _set_assert_marker ($hashed_string) {
```

File: t/03-testapi.t [hunk 5]
```diff
@@ -334,6 +334,35 @@ subtest 'script_run' => sub {
...
 
+sub assert_script_sudo_test () {
+    my $mock_testapi = Test::MockModule->new('testapi');
+    $mock_testapi->redefine(_handle_found_needle => sub { return $_[0] });
+    $mock_testapi->noop(qw(send_key enter_cmd));
+    $mock_testapi->redefine(hashed_string => 'XXX');
+    $mock_testapi->redefine(wait_serial => "XXX-0-");
+    my $script_sudo = '';
+    $mock_testapi->redefine(script_sudo => sub { $script_sudo = "$_[0]" });
+    $mock_testapi->redefine(_set_assert_marker => sub { return "marker"; });
```
Comment (okurz) [inline t/03-testapi.t:345]:
```suggestion
    $mock_testapi->redefine(_set_assert_marker => 'marker');
```

File: t/03-testapi.t [hunk 6]
```diff
@@ -334,6 +334,35 @@ subtest 'script_run' => sub {
     is(background_script_run('sleep 10', output => 'foo'), '1234', 'background_script_run with output returns valid PID');
 };
 
+sub assert_script_sudo_test () {
+    my $mock_testapi = Test::MockModule->new('testapi');
+    $mock_testapi->redefine(_handle_found_needle => sub { return $_[0] });
+    $mock_testapi->noop(qw(send_key enter_cmd));
+    $mock_testapi->redefine(hashed_string => 'XXX');
+    $mock_testapi->redefine(wait_serial => "XXX-0-");
```
Comment (okurz) [inline t/03-testapi.t:342]:
```suggestion
    $mock_testapi->redefine(wait_serial => 'XXX-0-');
```

File: t/03-testapi.t [hunk 7]
```diff
@@ -334,6 +334,35 @@ subtest 'script_run' => sub {
...
+    is assert_script_sudo('bash'), undef, 'successful assertion of script_sudo (2)';
+    is $script_sudo, 'bash; marker', 'script_sudo called like expected(2)';
+}
+
+sub set_assert_marker_test ($is_serial_terminal) {
+    my $mock_testapi = Test::MockModule->new('testapi');
+    $mock_testapi->redefine(_handle_found_needle => sub { return $_[0] });
+    $mock_testapi->redefine(is_serial_terminal => $is_serial_terminal);
+    my $expected_output = $is_serial_terminal ? ' > /dev/null' : '';
+    is testapi::_set_assert_marker('XXX'), "echo XXX-\$?-$expected_output", 'marker setter return correct string to the serial';
```
Comment (okurz) [inline t/03-testapi.t:357]:
```suggestion
    is testapi::_set_assert_marker('XXX'), "echo XXX-\$?-$expected_output", 'marker setter returns correct string to the serial';
```

**Review by okurz — CHANGES_REQUESTED** [review]

File: t/03-testapi.t (hunk 3 above)
Comment (b10n1k) [reply t/03-testapi.t]: *this should be ready for review*

File: testapi.pm (hunk 4 above)
Comment (b10n1k) [reply testapi.pm]: *ready for review*

File: t/03-testapi.t (hunk 5 above)
Comment (b10n1k) [reply t/03-testapi.t]: *ready for review*

File: t/03-testapi.t (hunk 6 above)
Comment (b10n1k) [reply t/03-testapi.t]: ady for review*

File: t/03-testapi.t (hunk 7 above)
Comment (b10n1k) [reply t/03-testapi.t]: *ready for review*

**Review by Martchus — APPROVED** [review]

**Comment by b10n1k** [comment]
@os-autoinst/tools-team Can this be merged?

**Comment by okurz** [comment]
> @os-autoinst/tools-team Can this be merged?

As soon as perlpunk approved, yes

**Review by kraih — APPROVED** [review]

**Review by perlpunk — APPROVED** [review]

**Comment by DimStar77** [comment]
This seems to have resulted in a few failures.
Reporter as https://progress.opensuse.org/issues/120984

**Pull request merged**